Debugger: Go build in the wrong directory

### Summary

<!-- Please insert a one line summary of the issue below -->
In a Zed project with two root directories, the go debugger tries to build in the first even if it is not the go project directory.

### Description
<!--  Describe with sufficient detail to reproduce from a clean Zed install. -->
Set up a project like this (or download [repro.zip](https://github.com/user-attachments/files/20697142/repro.zip)):

<img width="508" alt="Image" src="https://github.com/user-attachments/assets/e398a864-c674-400a-b117-3e0b34876cb7" />
<img width="738" alt="Image" src="https://github.com/user-attachments/assets/5a9065fb-4fe2-4183-b7c4-f8b58c768a5f" />

<details><summary>file contents (if you don't want to download a zip)</summary>

const.go
```go
package shared

const Text = "text"
```

main.go
```go
package main

import (
	"fmt"

	"github.com/smpanaro/go-servers/shared"
)

func main() {
	fmt.Println("Hello, World!", shared.Text)
}

```

mod.go
```go
module github.com/smpanaro/go-servers

go 1.24.2

```

package.json
```json
{}
```

debug.json
```jsonc
// Project-local debug tasks
//
// For more documentation on how to configure debug tasks,
// see: https://zed.dev/docs/debugger
[
  {
    "label": "app1-custom",
    "adapter": "Delve",
    "request": "launch",
    "mode": "debug",
    "program": "${ZED_WORKTREE_ROOT}/app1/main.go",
    "args": []
  }
]
```

</details>

Steps to trigger the problem:
1. Make sure that typescript-app is on the top in the sidebar (drag it up if not).
a. Also note that the `shared/` directory is required.
3. From the palette choose "debugger: start" then "app1-custom"<img width="496" alt="Image" src="https://github.com/user-attachments/assets/55776aaf-64e4-46c0-a400-faa1dc681428" />


**Expected Behavior**: The debugger starts.
**Actual Behavior**: The debugger errors with:
```
Build Error: go build -o /Users/me/Downloads/repro/typescript-app/__debug_bin2353077744 -gcflags all=-N -l /Users/me/Downloads/repro/go-servers/app1/main.go
../go-servers/app1/main.go:6:2: no required module provides package github.com/smpanaro/go-servers/shared: go.mod file not found in current directory or any parent directory; see 'go help modules' (exit status 1)
Tried to launch debugger with: {
  "request": "launch",
  "mode": "debug",
  "program": "/Users/me/Downloads/repro/go-servers/app1/main.go",
  "args": []
}
error: Failed to launch: Build error: Check the debug console for details.
```

**Workaround**:
Drag the go project above the typescript project.

### Zed Version and System Specs

Zed: v0.191.0 (Zed Preview) 
OS: macOS 15.5.0
Memory: 64 GiB
Architecture: aarch64

--- a/crates/debugger_ui/src/debugger_panel.rs
+++ b/crates/debugger_ui/src/debugger_panel.rs
@@ -179,6 +179,19 @@ impl DebugPanel {
                 cx,
             )
         });
+        let worktree = worktree_id.or_else(|| {
+            active_buffer
+                .as_ref()
+                .and_then(|buffer| buffer.read(cx).file())
+                .map(|f| f.worktree_id(cx))
+        });
+        let Some(worktree) = worktree
+            .and_then(|id| self.project.read(cx).worktree_for_id(id, cx))
+            .or_else(|| self.project.read(cx).visible_worktrees(cx).next())
+        else {
+            log::debug!("Could not find a worktree to spawn the debug session in");
+            return;
+        };
         self.debug_scenario_scheduled_last = true;
         if let Some(inventory) = self
             .project
@@ -213,7 +226,7 @@ impl DebugPanel {
                     .await?;
                 dap_store
                     .update(cx, |dap_store, cx| {
-                        dap_store.boot_session(session.clone(), definition, cx)
+                        dap_store.boot_session(session.clone(), definition, worktree, cx)
                     })?
                     .await
             }

--- a/crates/project/src/debugger/dap_store.rs
+++ b/crates/project/src/debugger/dap_store.rs
@@ -180,15 +180,12 @@ impl DapStore {
         &mut self,
         definition: DebugTaskDefinition,
         session_id: SessionId,
+        worktree: &Entity<Worktree>,
         console: UnboundedSender<String>,
         cx: &mut Context<Self>,
     ) -> Task<Result<DebugAdapterBinary>> {
         match &self.mode {
             DapStoreMode::Local(_) => {
-                let Some(worktree) = self.worktree_store.read(cx).visible_worktrees(cx).next()
-                else {
-                    return Task::ready(Err(anyhow!("Failed to find a worktree")));
-                };
                 let Some(adapter) = DapRegistry::global(cx).adapter(&definition.adapter) else {
                     return Task::ready(Err(anyhow!("Failed to find a debug adapter")));
                 };
@@ -229,6 +226,7 @@ impl DapStore {
                 let request = ssh.upstream_client.request(proto::GetDebugAdapterBinary {
                     session_id: session_id.to_proto(),
                     project_id: ssh.upstream_project_id,
+                    worktree_id: worktree.read(cx).id().to_proto(),
                     definition: Some(definition.to_proto()),
                 });
                 let ssh_client = ssh.ssh_client.clone();
@@ -401,12 +399,9 @@ impl DapStore {
         &self,
         session: Entity<Session>,
         definition: DebugTaskDefinition,
+        worktree: Entity<Worktree>,
         cx: &mut Context<Self>,
     ) -> Task<Result<()>> {
-        let Some(worktree) = self.worktree_store.read(cx).visible_worktrees(cx).next() else {
-            return Task::ready(Err(anyhow!("Failed to find a worktree")));
-        };
-
         let dap_store = cx.weak_entity();
         let console = session.update(cx, |session, cx| session.console_output(cx));
         let session_id = session.read(cx).session_id();
@@ -416,7 +411,13 @@ impl DapStore {
             async move |this, cx| {
                 let binary = this
                     .update(cx, |this, cx| {
-                        this.get_debug_adapter_binary(definition.clone(), session_id, console, cx)
+                        this.get_debug_adapter_binary(
+                            definition.clone(),
+                            session_id,
+                            &worktree,
+                            console,
+                            cx,
+                        )
                     })?
                     .await?;
                 session
@@ -780,9 +781,22 @@ impl DapStore {
         })
         .detach();
 
+        let worktree = this
+            .update(&mut cx, |this, cx| {
+                this.worktree_store
+                    .read(cx)
+                    .worktree_for_id(WorktreeId::from_proto(envelope.payload.worktree_id), cx)
+            })?
+            .context("Failed to find worktree with a given ID")?;
         let binary = this
             .update(&mut cx, |this, cx| {
-                this.get_debug_adapter_binary(definition, SessionId::from_proto(session_id), tx, cx)
+                this.get_debug_adapter_binary(
+                    definition,
+                    SessionId::from_proto(session_id),
+                    &worktree,
+                    tx,
+                    cx,
+                )
             })?
             .await?;
         Ok(binary.to_proto())

--- a/crates/proto/proto/debugger.proto
+++ b/crates/proto/proto/debugger.proto
@@ -493,6 +493,7 @@ message GetDebugAdapterBinary {
     uint64 project_id = 1;
     uint64 session_id = 3;
     DebugTaskDefinition definition = 2;
+    uint64 worktree_id = 4;
 }
 
 message DebugAdapterBinary {

--- a/crates/tasks_ui/src/tasks_ui.rs
+++ b/crates/tasks_ui/src/tasks_ui.rs
@@ -285,10 +285,12 @@ pub fn task_contexts(
                 .worktree_for_id(*worktree_id, cx)
                 .map_or(false, |worktree| is_visible_directory(&worktree, cx))
         })
-        .or(workspace
-            .visible_worktrees(cx)
-            .next()
-            .map(|tree| tree.read(cx).id()));
+        .or_else(|| {
+            workspace
+                .visible_worktrees(cx)
+                .next()
+                .map(|tree| tree.read(cx).id())
+        });
 
     let active_editor = active_item.and_then(|item| item.act_as::<Editor>(cx));
 

